vpdq/cpp/hashing/filehasher.cpp

```diff
@@ -62,322 +104,408 @@ static void saveFrameToFile(AVFrame* frame, const char* filename) {
+          AVMEDIA_TYPE_VIDEO) {
+        videoStreamIndex = i;
+        break;
+      }
     }
-  }
 
-  if (videoStreamIndex == -1) {
-    std::cerr << "No video stream found" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
-  }
+    if (videoStreamIndex == -1) {
+      avformat_close_input(&formatContext);
+      throw std::runtime_error("No video stream found");
+    }
 
-  // Get the video codec parameters
-  AVCodecParameters* codecParameters =
-      formatContext->streams[videoStreamIndex]->codecpar;
+    // Get the video codec parameters
+    AVCodecParameters* codecParameters =
+        formatContext->streams[videoStreamIndex]->codecpar;
 
-  // Get the width and height
-  // If downsampleWidth or downsampleHeight is 0,
-  // then use the video's original dimensions
-  int width = downsampleWidth;
-  int height = downsampleHeight;
-  if (width == 0) {
     width = codecParameters->width;
-  }
-  if (height == 0) {
     height = codecParameters->height;
-  }
+    if (width == 0 || height == 0) {
+      avformat_close_input(&formatContext);
+      throw std::runtime_error("Width or height equals 0");
+    }
 
-  if (width == 0 || height == 0) {
-    std::cerr << "Width or height equals 0" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
-  }
+    // Find the video decoder
+    const AVCodec* codec = avcodec_find_decoder(codecParameters->codec_id);
+    if (!codec) {
+      avformat_close_input(&formatContext);
+      throw std::runtime_error("Video codec id not found");
+    }
 
-  std::unique_ptr<vpdq::hashing::AbstractFrameBufferHasher> phasher =
-      vpdq::hashing::FrameBufferHasherFactory::createFrameHasher(height, width);
-  if (phasher == nullptr) {
-    std::cerr << "phasher allocation failed" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
-  }
+    // Create the codec context
+    codecContext = avcodec_alloc_context3(codec);
+    if (codecContext == nullptr) {
+      avformat_close_input(&formatContext);
+      throw std::bad_alloc();
+    }
+    if (avcodec_parameters_to_context(codecContext, codecParameters) < 0) {
+      avcodec_free_context(&codecContext);
+      avformat_close_input(&formatContext);
+      throw std::runtime_error("Cannot copy codec parameters to context");
+    }
 
-  // Find the video decoder
-  const AVCodec* codec = avcodec_find_decoder(codecParameters->codec_id);
-  if (!codec) {
-    std::cerr << "Codec decoder not found" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
+    // Determine the number of threads to use and multithreading type
+    codecContext->thread_count = 0;
+
+    if (codec->capabilities & AV_CODEC_CAP_FRAME_THREADS) {
+      codecContext->thread_type = FF_THREAD_FRAME;
+    } else if (codec->capabilities & AV_CODEC_CAP_SLICE_THREADS) {
+      codecContext->thread_type = FF_THREAD_SLICE;
+    } else {
+      codecContext->thread_count = 1;
```

Comment:
Dumb question - should we be passing the thread count passed in to the hash function to AV?

Reply:
I was on the fence. I'm pretty sure that for decoding it won't spawn a ton of threads because that's not very helpful for decoding, but if you want me to I can pass it. I haven't tried setting it to anything but 0 or 1 though so I would need to do some testing.

Reply:
I'm happy to leave as a future improvement if it doesn't seem like the answer is straightforward